### PR TITLE
Add e2s API

### PR DESCRIPTION
**Pull request by john0312**

### CODE REVIEW

File: common/extlib/extension-helper-standalone.mjs [hunk 1]
```diff
@@ -83,6 +84,24 @@ class ExtensionHelperStandalone extends ExtensionHelperBase {
     }
   }
 
+  /**
+   * This is called when an external service/client calls the E2s RPC API
+   * of this extension.
+   */
+  async onCallE2s(serviceName, methodName, args) {
+    const actualMethodName = ExtConst.E2S_RPC_FUNC_PREFIX() + methodName;
+    if (typeof this.ext[actualMethodName] != 'function') {
```
Comment (ktpss95112) [inline common/extlib/extension-helper-standalone.mjs:93]: Use `!==` to replace `!=`.

File: common/extlib/extension-manager.mjs [hunk 2]
```diff
@@ -262,6 +262,29 @@ class ExtensionManager {
     return await this.callC2sInStandalone(extName, methodName, args, playerID);
   }
 
+  /**
+   * This is called by gateway service when E2s APIs are called.
+   * It'll call the standalone E2s APIs directly.
+   */
+  async onE2sCalled(extName, apiName, args) {
+    if (!(typeof extName === 'string')) {
```
Comment (ktpss95112) [inline common/extlib/extension-manager.mjs:270]: `typeof extName !== 'string'`

File: common/extlib/extension-manager.mjs [hunk 3]
```diff
@@ -262,6 +262,29 @@ class ExtensionManager {
...
+      return {'error': 'extName not string'};
+    }
+    if (!(typeof apiName === 'string')) {
```
Comment (ktpss95112) [inline common/extlib/extension-manager.mjs:273]: same as above

File: common/extlib/extension-manager.mjs [hunk 4]
```diff
@@ -262,6 +262,29 @@ class ExtensionManager {
...
+      return {'error': 'apiName not string'};
+    }
+    if (!(typeof args === 'object') || !Array.isArray(args)) {
```
Comment (ktpss95112) [inline common/extlib/extension-manager.mjs:276]: same as above

File: common/extlib/extension-manager.mjs [hunk 5]
```diff
@@ -262,6 +262,29 @@ class ExtensionManager {
...
+      return {'error': 'args not array'};
+    }
+    const extService = await this.dir.getExtensionServiceName(extName);
+    if (typeof extService != 'string') {
```
Comment (ktpss95112) [inline common/extlib/extension-manager.mjs:280]: `!==`

File: services/gateway/gateway-server.mjs [hunk 6]
```diff
@@ -102,7 +102,7 @@ async function mainServer() {
   const broadcaster = new AllAreaBroadcaster(rpcDirectory, gameMap, gameState);
   const extensionManager = new ExtensionManager(rpcDirectory, broadcaster, gameMap, gameState);
   const gatewayService = new GatewayService(rpcDirectory, gameMap, authServer,
-    broadcaster, io, extensionManager, movementManager);
+    broadcaster, io, extensionManager, movementManager, app);
```
Comment (ktpss95112) [inline services/gateway/gateway-server.mjs:105]: What is `app`? Maybe consider using a readable name like `httpServerApp`?

**Comment by ktpss95112** [comment]
I didn't see the usage or examples of e2s API. Maybe write down "who calls what function by what means" in jsdoc.

File: extensions/iobj-lib/standalone.mjs [hunk 7]
```diff
@@ -247,6 +247,17 @@ class Standalone {
     return FSM_ERROR;
   }
 
+  /**
+   * Allow external party to set a variable for example.
+   */
+  async e2s_writeVar(varName, val, playerID, objID) {
+    // Example for triggering:
```
Comment (john0312) [inline extensions/iobj-lib/standalone.mjs:254]: The example is here.

**Pull request merged**